In [1]:
import pandas as pd
import numpy as np
import helper
import project_helper
import project_tests

In [62]:
# 2018-11-06 - 2014-12-01	
price_csv = "data/processed_btc_prices_ohlcvvw_2014-12-01_to_2018-11-11.csv"
df = pd.read_csv(price_csv, parse_dates=['date'], index_col=False)
print('Loaded Data')

Loaded Data


In [63]:
# df = df.drop(columns='Unnamed: 0')
cols_to_drop = ['Unnamed: 0',
                'Open',
                'High',
                'Low',
                'Volume_(BTC)',
                'Volume_(Currency)',
                'Weighted_Price']
df = df.drop(columns=cols_to_drop)

In [64]:
df.head()

date        Close
0 2014-12-01 335.00000000
1 2014-12-02 377.85714286
2 2014-12-03 377.71571429
3 2014-12-05 378.00000000
4 2014-12-08 375.00000000

In [65]:
df['ticker'] = 'BTC'

In [66]:
df.tail()

date         Close ticker
1403 2018-11-06 6436.19735417    BTC
1404 2018-11-07 6500.06674306    BTC
1405 2018-11-08 6435.30268056    BTC
1406 2018-11-09 6353.39325694    BTC
1407 2018-11-10 6359.55989483    BTC

In [86]:
# 2013-07-01 - 2017-06-30
df_quote_media = pd.read_csv('https://raw.githubusercontent.com/fatmirboja/trading-with-momentum/master/eod-quotemedia.csv', parse_dates=['date'], index_col=False)


In [95]:
df_quote_media.describe()

adj_close
count 490737.00000000
mean      75.10047207
std       75.43880376
min        1.59000000
25%       36.78242355
50%       57.49959261
75%       87.40000000
max     1011.34000000

In [92]:
start_date = '2014-12-01'
end_date = '2018-11-06'
mask = (df_quote_media['date'] > start_date) & (df_quote_media['date'] <= end_date)
df_quote_media_by_BTC_date = df_quote_media.loc[mask]


In [105]:
df_quote_media_by_BTC_date.head()

date ticker   adj_close
359 2014-12-02      A 40.43408931
360 2014-12-03      A 40.67488308
361 2014-12-04      A 40.71341008
362 2014-12-05      A 40.74230533
363 2014-12-08      A 39.98139703

In [106]:
close_quote_media = df_quote_media_by_BTC_date.reset_index().pivot(index='date', columns='ticker', values='adj_close')


In [107]:
#close_quote_media.head()
close_quote_media.tail()

ticker               A         AAL          AAP         AAPL        ABBV  \
date                                                                       
2017-06-26 58.57854478 48.36234805 121.52159207 143.57270901 70.35520945   
2017-06-27 58.22256443 48.08474540 121.69121741 141.51491885 70.01668424   
2017-06-28 58.73675827 48.82832394 116.45278767 143.58255490 70.52930812   
2017-06-29 58.27398382 49.19515602 115.79424221 141.46568942 70.10373358   
2017-06-30 58.77942143 49.88916265 116.33305213 141.80044954 70.13275003   

ticker             ABC         ABT          ACN         ADBE         ADI  ...  \
date                                                                      ...   
2017-06-26 94.31043377 47.95227368 120.09101209 144.96000000 78.92677572  ...   
2017-06-27 93.85848253 47.71697322 119.94376955 142.54000000 76.54633554  ...   
2017-06-28 94.69360982 47.53069368 121.46527575 143.81000000 77.58471685  ...   
2017-06-29 94.08445815 47.77579833 120.72906307 141.24000000 76.15449354  ...   
2017-06-30 92.87597984 47.65814810 121.40637874 141.44000000 76.21326984  ...   

ticker              XL        XLNX         XOM        XRAY         XRX  \
date                                                                     
2017-06-26 42.76892496 65.99587865 78.12543603 63.56673975 27.95461459   
2017-06-27 43.14151074 63.78164638 78.00041995 63.92391201 27.75350225   
2017-06-28 43.30819385 64.67321778 78.40431807 64.82428373 28.28980181   
2017-06-29 43.27877918 62.88027749 77.60613845 64.10898129 28.12560699   
2017-06-30 42.94541296 63.01744232 77.63498832 64.41695873 27.74892476   

ticker             XYL         YUM          ZBH        ZION         ZTS  
date                                                                     
2017-06-26 54.05749897 73.49934641 127.97264293 41.75554533 62.43009343  
2017-06-27 53.87954816 72.74212810 127.16946735 41.95278457 62.46990854  
2017-06-28 54.34419748 72.91914017 127.42727680 42.37684891 62.65903032  
2017-06-29 54.27499439 72.23075989 126.81250043 43.38276899 62.21111032  
2017-06-30 54.79896064 72.53561401 127.31820357 43.30387330 62.09166499  

[5 rows x 495 columns]

In [108]:
close_quote_media.reset_index()

ticker       date           A         AAL          AAP         AAPL  \
0      2014-12-02 40.43408931 46.31586800 147.51513567 107.65573562   
1      2014-12-03 40.67488308 46.68360787 151.05200180 108.87664163   
2      2014-12-04 40.71341008 48.04811631 152.25413889 108.46341191   
3      2014-12-05 40.74230533 49.36423792 153.17809549 108.00322425   
4      2014-12-08 39.98139703 49.18036799 152.68134463 105.56141223   
..            ...         ...         ...          ...          ...   
645    2017-06-26 58.57854478 48.36234805 121.52159207 143.57270901   
646    2017-06-27 58.22256443 48.08474540 121.69121741 141.51491885   
647    2017-06-28 58.73675827 48.82832394 116.45278767 143.58255490   
648    2017-06-29 58.27398382 49.19515602 115.79424221 141.46568942   
649    2017-06-30 58.77942143 49.88916265 116.33305213 141.80044954   

ticker        ABBV         ABC         ABT          ACN         ADBE  ...  \
0      61.12322689 87.16718413 41.81200708  80.15923224  73.47000000  ...   
1      60.44378579 87.11983125 42.23733930  80.66267473  73.18000000  ...   
2      61.38794420 87.08194894 42.06165860  80.49486057  73.03000000  ...   
3      61.51147894 87.32818393 41.88597790  80.35501543  72.40000000  ...   
4      61.25558554 87.65965410 42.26507836  79.21760832  72.05000000  ...   
..             ...         ...         ...          ...          ...  ...   
645    70.35520945 94.31043377 47.95227368 120.09101209 144.96000000  ...   
646    70.01668424 93.85848253 47.71697322 119.94376955 142.54000000  ...   
647    70.52930812 94.69360982 47.53069368 121.46527575 143.81000000  ...   
648    70.10373358 94.08445815 47.77579833 120.72906307 141.24000000  ...   
649    70.13275003 92.87597984 47.65814810 121.40637874 141.44000000  ...   

ticker          XL        XLNX         XOM        XRAY         XRX  \
0      32.75881654 41.99329292 82.98741395 54.13099272 33.97452436   
1      33.02725556 42.94664426 83.65702256 54.49330218 34.88280283   
2      33.11056422 42.99706188 83.14600547 54.45413359 34.85825476   
3      33.10130770 43.23081629 82.66142029 54.86050772 35.15283157   
4      33.60115966 42.99247846 80.79356471 54.48840611 34.51458183   
..             ...         ...         ...         ...         ...   
645    42.76892496 65.99587865 78.12543603 63.56673975 27.95461459   
646    43.14151074 63.78164638 78.00041995 63.92391201 27.75350225   
647    43.30819385 64.67321778 78.40431807 64.82428373 28.28980181   
648    43.27877918 62.88027749 77.60613845 64.10898129 28.12560699   
649    42.94541296 63.01744232 77.63498832 64.41695873 27.74892476   

ticker         XYL         YUM          ZBH        ZION         ZTS  
0      36.21994350 52.32612223 110.93069870 26.58100530 43.60058443  
1      37.04875336 51.84533090 110.17455326 26.89962978 43.73701898  
2      36.74390376 51.90542982 110.47507046 26.86100863 42.87943037  
3      36.76295686 52.28605628 111.44449019 27.61412102 42.85993972  
4      36.02941250 51.05736735 111.33785305 27.50791286 42.97688362  
..             ...         ...          ...         ...         ...  
645    54.05749897 73.49934641 127.97264293 41.75554533 62.43009343  
646    53.87954816 72.74212810 127.16946735 41.95278457 62.46990854  
647    54.34419748 72.91914017 127.42727680 42.37684891 62.65903032  
648    54.27499439 72.23075989 126.81250043 43.38276899 62.21111032  
649    54.79896064 72.53561401 127.31820357 43.30387330 62.09166499  

[650 rows x 496 columns]

In [109]:
close = df.reset_index().pivot(index='date', columns='ticker', values='Close')
# close = df.set_index('date')


In [115]:
close[close.duplicated(['date'])]

KeyError: Index(['date'], dtype='object')

## merge dfs


In [119]:
close_inner = close.merge(close_quote_media, on='date', how='inner', validate='1:1')
# close_left = close.join(close_quote_media, 'date')

In [121]:
close_inner.tail()


ticker               BTC           A         AAL          AAP         AAPL  \
date                                                                         
2017-06-26 2438.37924332 58.57854478 48.36234805 121.52159207 143.57270901   
2017-06-27 2407.38202939 58.22256443 48.08474540 121.69121741 141.51491885   
2017-06-28 2541.07432639 58.73675827 48.82832394 116.45278767 143.58255490   
2017-06-29 2554.11070139 58.27398382 49.19515602 115.79424221 141.46568942   
2017-06-30 2498.86145150 58.77942143 49.88916265 116.33305213 141.80044954   

ticker            ABBV         ABC         ABT          ACN         ADBE  ...  \
date                                                                      ...   
2017-06-26 70.35520945 94.31043377 47.95227368 120.09101209 144.96000000  ...   
2017-06-27 70.01668424 93.85848253 47.71697322 119.94376955 142.54000000  ...   
2017-06-28 70.52930812 94.69360982 47.53069368 121.46527575 143.81000000  ...   
2017-06-29 70.10373358 94.08445815 47.77579833 120.72906307 141.24000000  ...   
2017-06-30 70.13275003 92.87597984 47.65814810 121.40637874 141.44000000  ...   

ticker              XL        XLNX         XOM        XRAY         XRX  \
date                                                                     
2017-06-26 42.76892496 65.99587865 78.12543603 63.56673975 27.95461459   
2017-06-27 43.14151074 63.78164638 78.00041995 63.92391201 27.75350225   
2017-06-28 43.30819385 64.67321778 78.40431807 64.82428373 28.28980181   
2017-06-29 43.27877918 62.88027749 77.60613845 64.10898129 28.12560699   
2017-06-30 42.94541296 63.01744232 77.63498832 64.41695873 27.74892476   

ticker             XYL         YUM          ZBH        ZION         ZTS  
date                                                                     
2017-06-26 54.05749897 73.49934641 127.97264293 41.75554533 62.43009343  
2017-06-27 53.87954816 72.74212810 127.16946735 41.95278457 62.46990854  
2017-06-28 54.34419748 72.91914017 127.42727680 42.37684891 62.65903032  
2017-06-29 54.27499439 72.23075989 126.81250043 43.38276899 62.21111032  
2017-06-30 54.79896064 72.53561401 127.31820357 43.30387330 62.09166499  

[5 rows x 496 columns]

In [122]:
btc_ticker = 'BTC'

project_helper.plot_stock(close_inner[btc_ticker], '{} Stock'.format(btc_ticker))

## Resample Adjusted Prices

The trading signal you'll develop in this project does not need to be based on daily prices, for instance, you can use month-end prices to perform trading once a month. To do this, you must first resample the daily adjusted closing prices into monthly buckets, and select the last observation of each month.

Implement the `resample_prices` to resample `close_prices` at the sampling frequency of `freq`.

In [123]:
def resample_prices(close_prices, freq='M'):
    """
    Resample close prices for each ticker at specified frequency.
    
    Parameters
    ----------
    close_prices : DataFrame
        Close prices for each ticker and date
    freq : str
        What frequency to sample at
        For valid freq choices, see http://pandas.pydata.org/pandas-docs/stable/timeseries.html#offset-aliases
    
    Returns
    -------
    prices_resampled : DataFrame
        Resampled prices for each ticker and date
    """
    prices_resampled = close_prices.resample(rule=freq).last()    
    
    return prices_resampled

project_tests.test_resample_prices(resample_prices)

Tests Passed


### View Data
Let's apply this function to `close` and view the results.

In [124]:
monthly_close = resample_prices(close_inner)
project_helper.plot_resampled_prices(
    monthly_close.loc[:, btc_ticker],
    close_inner.loc[:, btc_ticker],
    '{} Stock - Close Vs Monthly Close'.format(btc_ticker))

## Compute Log Returns

Compute log returns ($R_t$) from prices ($P_t$) as your primary momentum indicator:

$$R_t = log_e(P_t) - log_e(P_{t-1})$$

Implement the `compute_log_returns` function below, such that it accepts a dataframe (like one returned by `resample_prices`), and produces a similar dataframe of log returns. Use Numpy's [log function](https://docs.scipy.org/doc/numpy/reference/generated/numpy.log.html) to help you calculate the log returns.

In [125]:
def compute_log_returns(prices):
    """
    Compute log returns for each ticker.
    
    Parameters
    ----------
    prices : DataFrame
        Prices for each ticker and date
    
    Returns
    -------
    log_returns : DataFrame
        Log returns for each ticker and date
    """
    log_returns = np.log(prices) - np.log(prices.shift(1))
    return log_returns

project_tests.test_compute_log_returns(compute_log_returns)

Tests Passed


### View Data
Using the same data returned from `resample_prices`, we'll generate the log returns.

In [126]:
monthly_close_returns = compute_log_returns(monthly_close)
project_helper.plot_returns(
    monthly_close_returns.loc[:, btc_ticker],
    'Log Returns of {} Stock (Monthly)'.format(btc_ticker))

## Shift Returns
Implement the `shift_returns` function to shift the log returns to the previous or future returns in the time series. For example, the parameter `shift_n` is 2 and `returns` is the following:

```
                           Returns
               A         B         C         D
2013-07-08     0.015     0.082     0.096     0.020     ...
2013-07-09     0.037     0.095     0.027     0.063     ...
2013-07-10     0.094     0.001     0.093     0.019     ...
2013-07-11     0.092     0.057     0.069     0.087     ...
...            ...       ...       ...       ...
```

the output of the `shift_returns` function would be:
```
                        Shift Returns
               A         B         C         D
2013-07-08     NaN       NaN       NaN       NaN       ...
2013-07-09     NaN       NaN       NaN       NaN       ...
2013-07-10     0.015     0.082     0.096     0.020     ...
2013-07-11     0.037     0.095     0.027     0.063     ...
...            ...       ...       ...       ...
```
Using the same `returns` data as above, the `shift_returns` function should generate the following with `shift_n` as -2:
```
                        Shift Returns
               A         B         C         D
2013-07-08     0.094     0.001     0.093     0.019     ...
2013-07-09     0.092     0.057     0.069     0.087     ...
...            ...       ...       ...       ...       ...
...            ...       ...       ...       ...       ...
...            NaN       NaN       NaN       NaN       ...
...            NaN       NaN       NaN       NaN       ...
```
_Note: The "..." represents data points we're not showing._

In [127]:
def shift_returns(returns, shift_n):
    """
    Generate shifted returns
    
    Parameters
    ----------
    returns : DataFrame
        Returns for each ticker and date
    shift_n : int
        Number of periods to move, can be positive or negative
    
    Returns
    -------
    shifted_returns : DataFrame
        Shifted returns for each ticker and date
    """
    shifted_returns = returns.shift(shift_n)

    return shifted_returns

project_tests.test_shift_returns(shift_returns)

Tests Passed


### View Data
Let's get the previous month's and next month's returns.

In [128]:
prev_returns = shift_returns(monthly_close_returns, 1)
lookahead_returns = shift_returns(monthly_close_returns, -1)

project_helper.plot_shifted_returns(
    prev_returns.loc[:, btc_ticker],
    monthly_close_returns.loc[:, btc_ticker],
    'Previous Returns of {} Stock'.format(btc_ticker))
project_helper.plot_shifted_returns(
    lookahead_returns.loc[:, btc_ticker],
    monthly_close_returns.loc[:, btc_ticker],
    'Lookahead Returns of {} Stock'.format(btc_ticker))

## Generate Trading Signal

A trading signal is a sequence of trading actions, or results that can be used to take trading actions. A common form is to produce a "long" and "short" portfolio of stocks on each date (e.g. end of each month, or whatever frequency you desire to trade at). This signal can be interpreted as rebalancing your portfolio on each of those dates, entering long ("buy") and short ("sell") positions as indicated.

Here's a strategy that we will try:
> For each month-end observation period, rank the stocks by _previous_ returns, from the highest to the lowest. Select the top performing stocks for the long portfolio, and the bottom performing stocks for the short portfolio.

Implement the `get_top_n` function to get the top performing stock for each month. Get the top performing stocks from `prev_returns` by assigning them a value of 1. For all other stocks, give them a value of 0. For example, using the following `prev_returns`:

```
                                     Previous Returns
               A         B         C         D         E         F         G
2013-07-08     0.015     0.082     0.096     0.020     0.075     0.043     0.074
2013-07-09     0.037     0.095     0.027     0.063     0.024     0.086     0.025
...            ...       ...       ...       ...       ...       ...       ...
```

The function `get_top_n` with `top_n` set to 3 should return the following:
```
                                     Previous Returns
               A         B         C         D         E         F         G
2013-07-08     0         1         1         0         1         0         0
2013-07-09     0         1         0         1         0         1         0
...            ...       ...       ...       ...       ...       ...       ...
```
*Note: You may have to use Panda's [`DataFrame.iterrows`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.iterrows.html) with [`Series.nlargest`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.Series.nlargest.html) in order to implement the function. This is one of those cases where creating a vecorization solution is too difficult.*

Additional trading signals:
    Simple Moving Average
Rate of Change
Relative Strength Index
Moving Average Convergence Divergence
Bollinger Bands
Fibonacci Retracements
https://www.value-at-risk.net/backtesting-value-at-risk/ (First understand backtesting)
Good backtesting: https://arxiv.org/pdf/1801.04112.pdf

In [129]:
def get_top_n(prev_returns, top_n):
    """
    Select the top performing stocks
    
    Parameters
    ----------
    prev_returns : DataFrame
        Previous shifted returns for each ticker and date
    top_n : int
        The number of top performing stocks to get
    
    Returns
    -------
    top_stocks : DataFrame
        Top stocks for each ticker and date marked with a 1
    """
    # TODO: Implement Function
    top_stocks = prev_returns.copy()
    top_stocks[:] = 0

    for date, returns in prev_returns.iterrows():
        top_stocks.loc[date, returns.nlargest(top_n).index] = 1
    return top_stocks.astype('int64')

project_tests.test_get_top_n(get_top_n)

Tests Passed


### View Data
We want to get the best performing and worst performing stocks. To get the best performing stocks, we'll use the `get_top_n` function. To get the worst performing stocks, we'll also use the `get_top_n` function. However, we pass in `-1*prev_returns` instead of just `prev_returns`. Multiplying by negative one will flip all the positive returns to negative and negative returns to positive. Thus, it will return the worst performing stocks.

In [130]:
top_bottom_n = 50

df_long = get_top_n(prev_returns, top_bottom_n)
df_short = get_top_n(-1*prev_returns, top_bottom_n)
project_helper.print_top(df_long, 'Longed Stocks')
project_helper.print_top(df_short, 'Shorted Stocks')

10 Most Longed Stocks:
AMD, BTC, NVDA, FCX, ATVI, NFLX, INCY, AMZN, WYNN, HP
10 Most Shorted Stocks:
CHK, MRO, RRC, FCX, CF, DVN, FTI, KORS, ARNC, VIAB


In [158]:
df_long = get_top_n(prev_returns, top_bottom_n)


In [159]:
print(df_long['BTC'])

date
2014-12-31    0
2015-01-31    0
2015-02-28    0
2015-03-31    0
2015-04-30    0
2015-05-31    0
2015-06-30    0
2015-07-31    1
2015-08-31    0
2015-09-30    0
2015-10-31    0
2015-11-30    1
2015-12-31    1
2016-01-31    1
2016-02-29    0
2016-03-31    1
2016-04-30    0
2016-05-31    0
2016-06-30    1
2016-07-31    1
2016-08-31    0
2016-09-30    0
2016-10-31    1
2016-11-30    1
2016-12-31    0
2017-01-31    1
2017-02-28    0
2017-03-31    1
2017-04-30    0
2017-05-31    1
2017-06-30    1
Freq: M, Name: BTC, dtype: int64


## Projected BTC Returns
It's now time to check if your trading signal has the potential to become profitable!


In [160]:
# get returns from long positions by masking with boolean mask of holdings
long_returns = lookahead_returns.mask(~df_long.astype(bool)).fillna(0)

In [161]:
long_returns

ticker             BTC           A         AAL         AAP        AAPL  \
date                                                                     
2014-12-31  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2015-01-31  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2015-02-28  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2015-03-31  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2015-04-30  0.00000000  0.00000000 -0.13068039  0.00000000  0.00000000   
2015-05-31  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2015-06-30  0.00000000  0.00000000  0.00000000  0.08953503  0.00000000   
2015-07-31 -0.21372739  0.00000000  0.00000000  0.00000000  0.00000000   
2015-08-31  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2015-09-30  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2015-10-31  0.00000000  0.00000000  0.00000000 -0.19834401  0.00000000   
2015-11-30  0.12875763  0.00000000  0.02607491  0.00000000  0.00000000   
2015-12-31 -0.12419388 -0.10480259  0.00000000  0.00000000  0.00000000   
2016-01-31  0.14248103  0.00000000  0.00000000  0.00000000  0.00000000   
2016-02-29  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2016-03-31  0.08910934  0.00000000  0.00000000  0.00000000  0.00000000   
2016-04-30  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2016-05-31  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2016-06-30 -0.00526355  0.08115189  0.00000000  0.00000000  0.00000000   
2016-07-31 -0.13853926  0.00000000  0.00000000  0.00000000  0.00000000   
2016-08-31  0.00000000  0.00000000  0.00850369  0.00000000  0.00000000   
2016-09-30  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2016-10-31  0.05298214  0.00000000  0.00000000  0.00000000 -0.02178228   
2016-11-30  0.25381823  0.00000000  0.00536885  0.00000000  0.00000000   
2016-12-31  0.00000000  0.00000000  0.00000000 -0.02927975  0.00000000   
2017-01-31  0.22410775  0.00000000  0.00000000  0.00000000  0.00000000   
2017-02-28  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2017-03-31  0.22516556  0.00000000  0.00000000  0.00000000 -0.00006961   
2017-04-30  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   
2017-05-31  0.09473289  0.00000000  0.00000000  0.00000000  0.00000000   
2017-06-30  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000   

ticker           ABBV         ABC         ABT        ACN        ADBE  ...  \
date                                                                  ...   
2014-12-31 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2015-01-31 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2015-02-28 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2015-03-31 0.00000000  0.00000000  0.00000000 0.00000000  0.02826855  ...   
2015-04-30 0.00000000 -0.01281413  0.00000000 0.00000000  0.00000000  ...   
2015-05-31 0.00897001  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2015-06-30 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2015-07-31 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2015-08-31 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2015-09-30 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2015-10-31 0.00000000  0.00000000  0.00000000 0.00000000  0.03109289  ...   
2015-11-30 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2015-12-31 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2016-01-31 0.00000000 -0.02946371  0.00000000 0.00000000  0.00000000  ...   
2016-02-29 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2016-03-31 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2016-04-30 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2016-05-31 0.00000000  0.00000000  0.00000000 0.00000000  0.00000000  ...   
2016-06-30 0.0000

In [162]:
long_returns['BTC'].sum()

0.7294304716448066

### View BTC Data
Time to see how btc did

In [168]:
# expected_portfolio_returns = portfolio_returns(df_long, df_short, lookahead_returns, 2*top_bottom_n)
project_helper.plot_returns(expected_portfolio_returns['BTC'], 'BTC Returns')

## Projected Returns
It's now time to check if your trading signal has the potential to become profitable!

We'll start by computing the net returns this portfolio would return. For simplicity, we'll assume every stock gets an equal dollar amount of investment. This makes it easier to compute a portfolio's returns as the simple arithmetic average of the individual stock returns.

Implement the `portfolio_returns` function to compute the expected portfolio returns. Using `df_long` to indicate which stocks to long and `df_short` to indicate which stocks to short, calculate the returns using `lookahead_returns`. To help with calculation, we've provided you with `n_stocks` as the number of stocks we're investing in a single period.

In [150]:
def portfolio_returns(df_long, df_short, lookahead_returns, n_stocks):
    """
    Compute expected returns for the portfolio, assuming equal investment in each long/short stock.
    
    Parameters
    ----------
    df_long : DataFrame
        Top stocks for each ticker and date marked with a 1
    df_short : DataFrame
        Bottom stocks for each ticker and date marked with a 1
    lookahead_returns : DataFrame
        Lookahead returns for each ticker and date
    n_stocks: int
        The number number of stocks chosen for each month
    
    Returns
    -------
    portfolio_returns : DataFrame
        Expected portfolio returns for each ticker and date
    """
    # get returns from long positions by masking with boolean mask of holdings
    long_returns = lookahead_returns.mask(~df_long.astype(bool)).fillna(0)
    # get returns from short positions by masking with boolean mask of holdings
    short_returns = lookahead_returns.mask(~df_short.astype(bool)).fillna(0)
    # do simple average
    portfolio_returns = (long_returns - short_returns)/n_stocks
    
    return portfolio_returns

project_tests.test_portfolio_returns(portfolio_returns)

Tests Passed


### View Data
Time to see how the portfolio did.

In [165]:
expected_portfolio_returns = portfolio_returns(df_long, df_short, lookahead_returns, 2*top_bottom_n)
project_helper.plot_returns(expected_portfolio_returns.T.sum(), 'Portfolio Returns')

In [167]:
expected_portfolio_returns['BTC']

date
2014-12-31    0.00000000
2015-01-31    0.00000000
2015-02-28    0.00031050
2015-03-31    0.00000000
2015-04-30    0.00000000
2015-05-31    0.00000000
2015-06-30    0.00000000
2015-07-31   -0.00213727
2015-08-31    0.00000000
2015-09-30   -0.00323903
2015-10-31    0.00000000
2015-11-30    0.00128758
2015-12-31   -0.00124194
2016-01-31    0.00142481
2016-02-29    0.00000000
2016-03-31    0.00089109
2016-04-30   -0.00160915
2016-05-31    0.00000000
2016-06-30   -0.00005264
2016-07-31   -0.00138539
2016-08-31    0.00000000
2016-09-30   -0.00145110
2016-10-31    0.00052982
2016-11-30    0.00253818
2016-12-31    0.00000000
2017-01-31    0.00224108
2017-02-28    0.00000000
2017-03-31    0.00225166
2017-04-30   -0.00528729
2017-05-31    0.00094733
2017-06-30    0.00000000
Freq: M, Name: BTC, dtype: float64

## Statistical Tests
### Annualized Rate of Return

In [138]:
expected_portfolio_returns_by_date = expected_portfolio_returns.T.sum().dropna()
portfolio_ret_mean = expected_portfolio_returns_by_date.mean()
portfolio_ret_ste = expected_portfolio_returns_by_date.sem()
portfolio_ret_annual_rate = (np.exp(portfolio_ret_mean * 12) - 1) * 100

print("""
Mean:                       {:.6f}
Standard Error:             {:.6f}
Annualized Rate of Return:  {:.2f}%
""".format(portfolio_ret_mean, portfolio_ret_ste, portfolio_ret_annual_rate))


Mean:                       0.003831
Standard Error:             0.002949
Annualized Rate of Return:  4.70%



The annualized rate of return allows you to compare the rate of return from this strategy to other quoted rates of return, which are usually quoted on an annual basis. 

### T-Test
Our null hypothesis ($H_0$) is that the actual mean return from the signal is zero. We'll perform a one-sample, one-sided t-test on the observed mean return, to see if we can reject $H_0$.

We'll need to first compute the t-statistic, and then find its corresponding p-value. The p-value will indicate the probability of observing a t-statistic equally or more extreme than the one we observed if the null hypothesis were true. A small p-value means that the chance of observing the t-statistic we observed under the null hypothesis is small, and thus casts doubt on the null hypothesis. It's good practice to set a desired level of significance or alpha ($\alpha$) _before_ computing the p-value, and then reject the null hypothesis if $p < \alpha$.

For this project, we'll use $\alpha = 0.05$, since it's a common value to use.

Implement the `analyze_alpha` function to perform a t-test on the sample of portfolio returns. We've imported the `scipy.stats` module for you to perform the t-test.

Note: [`scipy.stats.ttest_1samp`](https://docs.scipy.org/doc/scipy-1.0.0/reference/generated/scipy.stats.ttest_1samp.html) performs a two-sided test, so divide the p-value by 2 to get 1-sided p-value

In [139]:
from scipy import stats

def analyze_alpha(expected_portfolio_returns_by_date):
    """
    Perform a t-test with the null hypothesis being that the expected mean return is zero.
    
    Parameters
    ----------
    expected_portfolio_returns_by_date : Pandas Series
        Expected portfolio returns for each date
    
    Returns
    -------
    t_value
        T-statistic from t-test
    p_value
        Corresponding p-value
    """
    t_value, p_value = stats.ttest_1samp(expected_portfolio_returns_by_date, popmean=0)

    return t_value, p_value/2.0

project_tests.test_analyze_alpha(analyze_alpha)

Tests Passed


### View Data
Let's see what values we get with our portfolio. After you run this, make sure to answer the question below.

In [140]:
t_value, p_value = analyze_alpha(expected_portfolio_returns_by_date)
print("""
Alpha analysis:
 t-value:        {:.3f}
 p-value:        {:.6f}
""".format(t_value, p_value))


Alpha analysis:
 t-value:        1.299
 p-value:        0.101943



### What p-value did you observe? And what does that indicate about your signal?

A p-value of 0.101943 is greater than 0.05, thus we cannot reject the null hypothesis at alpha=0.05 level.  Overall, this supports the claim that our results are not statistically significant.